src/client/pfs_v2.go

```diff
@@ -50,57 +51,72 @@ func (c APIClient) NewPutTarClient(repo, commit string) (_ *PutTarClient, retErr
 // PutTar puts a tar stream into PFS.
 // The files are not persisted until the PutTarClient is closed.
 func (ptc *PutTarClient) PutTar(r io.Reader, tag ...string) error {
-	if len(tag) > 0 {
-		if len(tag) > 1 {
-			return errors.Errorf("PutTar called with %v tags, expected 0 or 1", len(tag))
+	return ptc.maybeError(func() error {
+		if len(tag) > 0 {
+			if len(tag) > 1 {
+				return errors.Errorf("PutTar called with %v tags, expected 0 or 1", len(tag))
+			}
+			if err := ptc.client.Send(&pfs.PutTarRequestV2{Tag: tag[0]}); err != nil {
+				return err
+			}
 		}
-		if err := ptc.client.Send(&pfs.PutTarRequest{Tag: tag[0]}); err != nil {
+		if _, err := grpcutil.ChunkReader(r, func(data []byte) error {
+			return ptc.client.Send(&pfs.PutTarRequestV2{Data: data})
+		}); err != nil {
 			return err
 		}
+		return ptc.client.Send(&pfs.PutTarRequestV2{EOF: true})
+	})
+}
+
+func (ptc *PutTarClient) maybeError(f func() error) (retErr error) {
+	if ptc.err != nil {
+		return ptc.err
 	}
-	if _, err := grpcutil.ChunkReader(r, func(data []byte) error {
-		return ptc.client.Send(&pfs.PutTarRequest{Data: data})
-	}); err != nil {
-		return err
-	}
-	return ptc.client.Send(&pfs.PutTarRequest{EOF: true})
+	defer func() {
+		retErr = grpcutil.ScrubGRPC(retErr)
+		if retErr != nil {
+			ptc.err = retErr
+		}
+	}()
+	return f()
 }
 
 // Close closes the PutTarClient, which persists the files.
 func (ptc *PutTarClient) Close() error {
-	_, err := ptc.client.CloseAndRecv()
-	return err
+	return ptc.maybeError(func() error {
+		_, err := ptc.client.CloseAndRecv()
+		return err
+	})
```

Comment:
Because `maybeError` doesn't call `f` if it's already encountered an error doesn't this mean there's a chance that a `Close` call doesn't get through to the underlying GRPC layer? I'm not sure quite what the semantics are, maybe you don't need to call `Close` if it encounters an earlier error? That would be a little unexpected to me, I generally assume that you need to call `Close` no matter what, and in some error cases it might be unnecessary, but shouldn't cause any problems if you do.

Reply:
We don't want to close if something got borked. That could result with incomplete data being persisted. We would need to figure out how to properly recover (which I don't think is worth it right now) before we would be safe doing this.

Reply:
In the case of PutTar, it seems like if something goes wrong we can just exit.  The user will need to inspect the commit, and see if they need to retry.  But the general case for GRPC is that you could have manipulated the server in any particular way with the messages leading up until now.  This case is basically just a HTTP POST that got interrupted.  It should be fine to give up without too much thought.